Developer-instructions.md

```diff
@@ -0,0 +1,253 @@
+root (i.e. inside the directory `...some-path.../coq-waterproof/`), run
+```
+dune build -p coq-waterproof
+```
+To then install the library so it is available from other projects, you can run
+```
+dune install -p coq-waterproof
+```
+
+## Testing
+
+We aim to test every feature that we implement. Currently we test almost
+all functionality present in `.v` files. Note for instance that tactics
+such as `Choose`, `Assume`, both have a file that defines them, and a file
+that tests them.
+
+## Running external tests
+
+If one has a folder `waterproof-exercises` at the same level as the folder
+`coq-waterproof`, one can test coq-waterproof against all the files 
+(typically exercise sheets) in `waterproof-exercises` by executing
+
+```
+dune build -p coq-waterproof @runtest
+```
+
+## Building with make
+
+In principle, we use **dune** to build `coq-waterproof`, but we try to 
+still support building with **make** as well. Before creating a PR, it is
+recommended to check that this really works. For this, both `_CoqProjectForMake`
+and `waterproof.mlpack` need to be in order, i.e. they need to contain
+the correct files, and in particular `waterproof.mlpack` needs the files 
+to be in the right order.
+
+TODO: include a test that checks that all .v, .ml files are actually included in the
+`_CoqProjectForMake` file.
+
+To then compile and install, run from the project's root directory:
+```
+make
+make install
+```
+
+## Building with make (legacy method)
+
+In previous versions of coq-waterproof, before running make, one needed to run,
+from the project's root directory,
+
+```
+autoreconf -i -s
+./configure
+```
+after which one could run
+```
+make
+make install
+```
+
+One needs to have the `autoconf` ocaml libary installed. One can install this with
+```
+sudo apt-get update
+sudo apt-get install autoconf
+```
+
+## Making Ocaml functions available from Ltac2: Using the foreign function interface (ffi)
+
+One can make Ocaml functions available for use in Ltac2 by using the
+foreign function interface (ffi). For now, we gather all such 
+definitions in `Wp_ffi.ml`.
+
+Suppose one has an Ocaml function `f : a -> b -> c -> d` where `a, b, c, d` are types, and one would like to make this function available as an Ltac2 tactic.
+
+### Make a tactic
+
+First of all, only tactics can pass the ffi, so one needs to convert `f` into
+something that outputs a tactic. An element of type `'a` can be converted into an `'a`-valued tactic, ie.e. an element of type `'a tactic` by using `tclUNIT`:
+
+Consider this example:
+```
+let my_unit_tactic : unit tactic = tclUNIT ()
+```
+
+```
+let my_bool_tactic : bool tactic = tclUNIT true
+```
+
+Hence, we can convert `f` into something that outputs a tactic by
+
+```
+let my_f_tactic : a -> b -> c -> d tactic =
+  fun input_a input_b input_c -> tclUNIT (f input_a input_b input_c)
+```
+
+### Make a `valexpr` tactic
+
+Yet datatypes in Ocaml and in Ltac2 need to be translated to each other
```

Comment:
```suggestion
Datatypes in Ocaml and in Ltac2 need to be translated to each other
```
If this was into contrast to the part before I don't really understand it, so alternative to removing yet is explaining the contrast in a bit more depth.